### PR TITLE
[20120] TCPSendResources cleanup

### LINKED ISSUE #4122: send_resource in participant not delete when participant delete
### Is there an already existing issue for this?

- [X] I have searched the existing issues

### Expected behavior

When ClientParticipant delete rightly, the ClientParticipant's related `send_source` is deleted rightly so that there will be no warning info.

### Current behavior

Using TCPTransport, when ClientParticipant delete rightly, the ClientParticipant's related `send_source` is still in ServerParticipant's  `send_source_list_`, which caused warning info.
Not connect
send_resource_list_size

### Steps to reproduce

create a ClientParticipant, connect to ServerParticipant, delete ClientParticipant
create a ClientParticipant, connect to ServerParticipant again.

### Fast DDS version/commit

2.11.2

### Platform/Architecture

Ubuntu Focal 20.04 amd64

### Transport layer

TCPv4

### Additional context

_No response_

### XML configuration file

_No response_

### Relevant log output

_No response_

### Network traffic capture

_No response_

### PR DESCRIPTION
**Pull request by jepemi**
## Description

TCP send resources weren't being cleaned after the disconnection of the other peer. This lead to a constantly increasing list of send resources which had to be traversed on each send call. 

This PR introduces a new callback that erases the send resource associated to a remote participant that has been removed, except if it belongs to the initial peers list. It also erases the channel from the channel resources map of the associated TCPTransport.

@Mergifyio backport 2.12.x 2.10.x 2.6.x

 Fixes #4122 

- This PR depends on #4277 and must be merged after that one.

## Contributor Checklist
- [x] Commit messages follow the project guidelines. 
- [x] The code follows the style guidelines of this project. 
- [x] Tests that thoroughly check the new feature have been added/Regression tests checking the bug and its fix have been added; the added tests pass locally 
- [x] Any new/modified methods have been properly documented using Doxygen. 
- [x] Changes are ABI compatible. 
- [x] Changes are API compatible. 
- N/A New feature has been added to the `versions.md` file (if applicable).
- N/A New feature has been documented/Current behavior is correctly described in the documentation. 
    
- [x] Applicable backports have been included in the description.

## Reviewer Checklist
- [x] The PR has a milestone assigned.
- [x] The title and description correctly express the PR's purpose.
- [x] Check contributor checklist is correct.
- [x] Check CI results: changes do not issue any warning.
- [x] Check CI results: failing tests are unrelated with the changes.

### CODE REVIEW

**Comment by EduPonz** [comment]
@Mergifyio backport 2.12.x 2.11.x 2.10.x 2.6.x

**Comment by jepemi** [comment]
@richiprosima please test this

**Comment by JesusPoderoso** [comment]
@jepemi can you address the conflicts, please? 😀

**Comment by JesusPoderoso** [comment]
Implementation would not be ready for the v2.13.3 release. I'm moving this to 2.13.4

**Comment by JesusPoderoso** [comment]
@richiprosima please test this

File: test/unittest/transport/TCPv4Tests.cpp [hunk 1]
```diff
@@ -2164,6 +2164,41 @@ TEST_F(TCPv4Tests, opening_output_channel_with_same_locator_as_local_listening_p
...
+    sendTransportUnderTest.init();
+
+    Locator_t outputLocator;
+    IPLocator::createLocator(LOCATOR_KIND_TCPv4, "127.0.0.1", g_default_port, outputLocator);
+    IPLocator::setLogicalPort(outputLocator, 7410);
+
+    SendResourceList send_resource_list;
+    ASSERT_TRUE(sendTransportUnderTest.OpenOutputChannel(send_resource_list, outputLocator));
+
+    std::this_thread::sleep_for(std::chrono::milliseconds(100));
```
[the remark this answers is not in the document]
Comment (jepemi) [reply test/unittest/transport/TCPv4Tests.cpp:2187]: At the end of the test, if the participant was destroyed while trying to connect to the receiving transport, asio would throw: `local_endpoint: Bad file descriptor`.
By removing the receiving transport (which, in fact, is not necessary), there is no need to wait.
Removing it from v6 test too.

File: src/cpp/rtps/participant/RTPSParticipantImpl.h [hunk 2]
```diff
@@ -1266,9 +1267,13 @@ class RTPSParticipantImpl
...
-    //! Remove participants from send resource list
-    void remove_from_send_resource_list(
-            std::set<Locator_t>& remote_participant_physical_locators);
+    /**
+     * Method called on participant removal with the set of locators associated to the participant.
+     *
+     * @param remote_participant_locators Set of locators associated to the participant removed.
+     */
+    void update_removed_participant(
+            LocatorList_t& remote_participant_locators);
```
Comment (EduPonz) [inline src/cpp/rtps/participant/RTPSParticipantImpl.h:1276]:
```suggestion
            const LocatorList_t& remote_participant_locators);
```

File: test/unittest/transport/test_UDPv4Tests.cpp
Comment (EduPonz) [inline test/unittest/transport/test_UDPv4Tests.cpp:1]: I think we can remove this file, can't we?

File: test/unittest/transport/TCPv4Tests.cpp [hunk 3]
```diff
@@ -1200,7 +1201,7 @@ TEST_F(TCPv4Tests, send_and_receive_between_secure_clients_1)
         senderThread->join();
         sem.wait();
     }
-    ASSERT_TRUE(sendTransportUnderTest2.CloseOutputChannel(outputLocator));
+    ASSERT_TRUE(sendTransportUnderTest2.SenderResourceHasBeenClosed(outputLocator));
```
Comment (EduPonz) [inline test/unittest/transport/TCPv4Tests.cpp:1204]: Please change all the commented tests for TODOs

File: src/cpp/rtps/network/NetworkFactory.cpp [hunk 4]
```diff
@@ -471,20 +472,20 @@ void NetworkFactory::update_network_interfaces()
...
         SendResourceList& send_resource_list,
-        std::set<Locator_t>& remote_participant_physical_locators) const
+        const LocatorList_t& remote_participant_locators,
+        const LocatorList_t& participant_initial_peers) const
 {
     for (auto& transport : mRegisteredTransports)
     {
-        if (transport->kind() == LOCATOR_KIND_TCPv4 ||
-                transport->kind() == LOCATOR_KIND_TCPv6)
+        TCPTransportInterface* tcp_transport = dynamic_cast<TCPTransportInterface*>(transport.get());
```
[the remark this answers is not in the document]
Comment (EduPonz) [reply src/cpp/rtps/network/NetworkFactory.cpp:482]: 2a5ad9c

**Review by EduPonz — APPROVED** [review]

**Comment by EduPonz** [comment]
@richiprosima please test this

**Comment by JesusPoderoso** [comment]
@Mergifyio backport 2.13.x

**Pull request merged**